### PR TITLE
fix: vertex-anthropic compat — empty text blocks, tool_use SSE, and model visibility

### PR DESCRIPTION
**Pull request by cidrblock**
## Summary

- Fix Vertex Anthropic compatibility issues discovered after the engine was merged:
  - Sanitize empty/whitespace text content blocks from requests (Vertex rejects with 400)
  - Handle `tool_use` blocks in JSON-to-SSE conversion (previously failed with `non-text-content`)
  - Fall back to engine's static `supportsTools` metadata when model discovery returns nothing (VS Code was hiding models)
  - Use nullish coalescing (`??`) for message content in `web/server.ts`
- Update tests to match new tool_use SSE behavior and add coverage for non-Anthropic JSON payloads

## Commits

- `fix: vertex-anthropic empty text blocks, tool_use SSE, and model visibility` -- compat fixes for edge cases found during real Vertex Anthropic usage
- `test(core): update convertAnthropicJsonToSse tests for tool_use SSE support` -- align test expectations, add parse-error coverage
- `fix: address Copilot review feedback on vertex-anthropic compat` -- remove inaccurate comments, drop hardcoded supportsVision:false, fix empty-array edge cases in sanitizer and convertMessages, remove duplicate test, fix tautological assertion, update JSDoc/debug messages

## Test plan

- [x] `npm run lint` passes locally (0 errors)
- [x] `npm test` passes locally (420 tests, all green)
- [x] Tests updated for changed behavior
- [x] All Copilot review comments addressed and threads resolved

### CODE REVIEW

File: packages/daemon/src/daemon/web/server.ts [hunk 1]
```diff
@@ -540,7 +540,9 @@ export function createWebApp(state: DaemonState): Express {
     // Convert web messages to the format state.chat() expects
     const chatMessages = messages.map((m: { role?: string; content?: string; name?: string; tool_call_id?: string; tool_calls?: unknown[] }) => ({
       role: m.role || 'user',
-      content: m.content || '',
+      // Preserve null/undefined as-is so convertMessages can skip empty content
+      // blocks rather than producing empty strings that Vertex Anthropic rejects
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/daemon/web/server.ts]: Good catch — the comment was inaccurate since `??` still coerces to empty string. Removed the misleading comment in a37e637.

File: packages/daemon/src/core/state.ts [hunk 2]
```diff
@@ -462,14 +462,22 @@ export class CoreState {
         const compositeId = `${providerId}/${modelName}`;
         const discovered = discoveryMap.get(engineModelId);
 
+        // Fall back to engine's static metadata when discovery returns nothing.
+        // Without this, engines that have no discovery API (e.g. vertex-anthropic)
+        // would report supportsTools=false, causing VS Code/Copilot to hide the model.
+        const capabilities = discovered?.capabilities ?? {
+          supportsTools: engineInfo.supportsTools,
+          supportsVision: false,
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/core/state.ts]: Agreed. Removed the hardcoded `supportsVision: false` — the field is now omitted from the fallback so it stays `undefined` rather than incorrectly marking vision-capable models. Fixed in a37e637.

File: packages/daemon/src/core/engines.ts [hunk 3]
```diff
@@ -160,6 +162,50 @@ export function sanitizeVertexRequestBody(bodyStr: string): { body: string; remo
...
+      // Array content: filter whitespace-only text blocks
+      if (!Array.isArray(msg.content)) {
+        filteredMessages.push(msg);
+        continue;
+      }
+      const filtered = (msg.content as Array<Record<string, unknown>>).filter(
+        block => !(block.type === 'text' && typeof block.text === 'string' && block.text.trim() === ''),
+      );
+      if (filtered.length === 0 && (msg.content as unknown[]).length > 0) {
+        // All content was empty text blocks — drop the whole message
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/core/engines.ts]: Fixed — the sanitizer now drops messages when `filtered.length === 0` regardless of original array length, so empty arrays (including `content: []`) are properly caught. Also updated `convertMessages` to fall back to string content instead of emitting an empty array. Fixed in a37e637.

File: packages/daemon/src/core/engines.ts [hunk 4]
```diff
@@ -957,12 +1016,14 @@ function convertMessages(messages: ChatMessage[]): ModelMessage[] {
       case 'user':
         return { role: 'user' as const, content: m.content };
 
-      case 'assistant':
+      case 'assistant': {
+        // Always use array form so empty content is handled uniformly.
+        // sanitizeVertexRequestBody will drop messages with empty content arrays.
+        const parts: AssistantModelMessage['content'] = [];
+        if (m.content && m.content.trim()) {
+          parts.push({ type: 'text', text: m.content });
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/core/engines.ts]: Fixed both issues: `convertMessages` now returns string content when there are no parts (no more `content: []`), and the sanitizer also catches empty arrays as a safety net. Stale comment removed. Fixed in a37e637.

File: packages/daemon/src/core/engines.test.ts [hunk 5]
```diff
@@ -139,6 +146,13 @@ describe('convertAnthropicJsonToSse', () => {
     expect(convertAnthropicJsonToSse(randomObj)).toEqual({ ok: false, reason: 'parse-error' });
   });
 
+  it('should return parse-error for JSON that is not an Anthropic Messages response', () => {
+    const errorPayload = JSON.stringify({ error: { type: 'server_error', message: 'Internal error' } });
+    expect(convertAnthropicJsonToSse(errorPayload)).toEqual({ ok: false, reason: 'parse-error' });
+    const randomObj = JSON.stringify({ status: 'ok', data: [1, 2, 3] });
+    expect(convertAnthropicJsonToSse(randomObj)).toEqual({ ok: false, reason: 'parse-error' });
+  });
+
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/core/engines.test.ts]: Removed the duplicate test case. Fixed in a37e637.

File: packages/daemon/src/core/engines.test.ts [hunk 6]
```diff
@@ -153,15 +167,17 @@ describe('convertAnthropicJsonToSse', () => {
...
     const json = JSON.stringify({
       id: 'msg',
       content: [{ type: 'text', text: 'Hello ' }, { type: 'text', text: 'world' }],
     });
     const result = convertAnthropicJsonToSse(json);
     expect(result.ok).toBe(true);
     if (!result.ok) return;
-    expect(result.body).toContain('Hello world');
+    expect(result.body).toContain('Hello ');
+    expect(result.body).toContain('');
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply packages/daemon/src/core/engines.test.ts]: Replaced the empty-string assertion with a `content_block_delta` event count check (expects 2 for two text blocks). Fixed in a37e637.
